### PR TITLE
Add activity store

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: core/collection/task_syncer.go [hunk 1]
```diff
@@ -49,6 +49,44 @@ func (t TaskSyncer) UpdateAndSyncTask(task entity.Task) error {
...
+		},
+		Payload: realtime.UserDraggingTaskPayload{
+			UserID:   userID,
+			TaskID:   taskID,
+			ClientID: clientID,
+			TeamID:   owningTeamID,
+			Dragging: true,
+		},
+	})
+
```
Comment (naturetime10) [inline core/collection/task_syncer.go:67]:
```suggestion
```

File: core/collection/task_syncer.go [hunk 2]
```diff
@@ -49,6 +49,44 @@ func (t TaskSyncer) UpdateAndSyncTask(task entity.Task) error {
 	return nil
 }
 
+func (t TaskSyncer) StartDraggingSyncTask(taskID uint64, userID uint64, clientID uint64, owningTeamID uint64) error {
```
Comment (naturetime10) [inline core/collection/task_syncer.go:52]:
```suggestion
func (t TaskSyncer) StartDraggingSyncTask(taskID uint64, userID uint64, clientID uint64, owningTeamID uint64) {
```

File: core/collection/task_syncer.go [hunk 3]
```diff
@@ -49,6 +49,44 @@ func (t TaskSyncer) UpdateAndSyncTask(task entity.Task) error {
...
+			ClientID: clientID,
+			TeamID:   owningTeamID,
+			Dragging: true,
+		},
+	})
+
+	return nil
+}
+
+func (t TaskSyncer) StopDraggingSyncTask(taskID uint64, userID uint64, clientID uint64, owningTeamID uint64) error {
```
Comment (naturetime10) [inline core/collection/task_syncer.go:71]:
```suggestion
func (t TaskSyncer) StopDraggingSyncTask(taskID uint64, userID uint64, clientID uint64, owningTeamID uint64) {
```

File: core/collection/task_syncer.go [hunk 4]
```diff
@@ -49,6 +49,44 @@ func (t TaskSyncer) UpdateAndSyncTask(task entity.Task) error {
...
+		CollectionType: realtime.UserActivityCollectionType,
+		MutationType:   realtime.UpdateMutationType,
+		TeamIDs: []uint64{
+			owningTeamID,
+		},
+		Payload: realtime.UserDraggingTaskPayload{
+			UserID:   userID,
+			TaskID:   taskID,
+			ClientID: clientID,
+			TeamID:   owningTeamID,
```
Comment (naturetime10) [inline core/collection/task_syncer.go:63]:
```suggestion
```

File: core/entity/user_status.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package entity
+
+type UserStatus string
+
+const (
+	UserStatusInitialized  UserStatus = "Initialized"
+	UserStatusConnected    UserStatus = "Connected"
+	UserStatusDisConnected UserStatus = "DisConnected"
```
Comment (naturetime10) [inline core/entity/user_status.go:8]:
```suggestion
	DisconnectedUserStatus UserStatus = "Disconnected"
```

File: core/entity/user_status.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package entity
+
+type UserStatus string
+
+const (
+	UserStatusInitialized  UserStatus = "Initialized"
```
Comment (naturetime10) [inline core/entity/user_status.go:6]:
```suggestion
	InitializedUserStatus  UserStatus = "Initialized"
```

File: core/entity/user_status.go [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+	UserStatusConnected    UserStatus = "Connected"
```
Comment (naturetime10) [inline core/entity/user_status.go:7]:
```suggestion
	ConnectedUserStatus   UserStatus = "Connected"
```

File: core/realtime/activity_store.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		userActivity = a.newUserActivity(userID)
+	}
+
+	userActivity.state = entity.UserStatusConnected
+	return userActivity
+}
+
+func (a *ActivityStore) StartDraggingTask(userID uint64, clientID uint64, taskID uint64, teamID uint64) {
+	userActivity := a.getUserActivity(userID)
+
```
Comment (naturetime10) [inline core/realtime/activity_store.go:57]:
```suggestion
```

File: core/realtime/activity_store.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	teamActivity := userActivity.getTeamActivity(teamID)
+	clientActivity := teamActivity.getClientActivity(clientID)
+	clientActivity.taskAction.taskID = taskID
+	clientActivity.taskAction.dragging = true
+}
+
+func (a *ActivityStore) StopDraggingTask(userID uint64, clientID uint64, taskID uint64, teamID uint64) {
+	userActivity := a.getUserActivity(userID)
+
```
Comment (naturetime10) [inline core/realtime/activity_store.go:66]:
```suggestion
```

File: core/realtime/activity_store.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	return userActivity
+}
+
+func (a *ActivityStore) StartDraggingTask(userID uint64, clientID uint64, taskID uint64, teamID uint64) {
+	userActivity := a.getUserActivity(userID)
+
+	teamActivity := userActivity.getTeamActivity(teamID)
+	clientActivity := teamActivity.getClientActivity(clientID)
+	clientActivity.taskAction.taskID = taskID
+	clientActivity.taskAction.dragging = true
```
Comment (naturetime10) [inline core/realtime/activity_store.go:61]: I wonder whether we need clientActivity. Are user activity combined with team activity not enough?

File: core/realtime/client_activity.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+type ClientActivity struct {
+	clientID   uint64
+	taskAction *TaskAction
+}
+
+func newClientActivity(clientID uint64) *ClientActivity {
+	clientActivity := ClientActivity{
+		clientID:   clientID,
+		taskAction: &TaskAction{},
```
Comment (naturetime10) [inline core/realtime/client_activity.go:16]:
```suggestion
```

Let's leave it as nil by default

File: core/realtime/mutation.go [hunk 12]
```diff
@@ -17,3 +17,11 @@ type DeleteSprintTaskRelationPayload struct {
 	SprintID uint64
 	TaskID   uint64
 }
+
+type UserDraggingTaskPayload struct {
```
Comment (naturetime10) [inline core/realtime/mutation.go:21]:
```suggestion
type TeamActivityPayload struct {
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/realtime/activity_store.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		if len(teamActivity.clientActivities) == 0 {
+			delete(a.userActivities[userID].teamActivities, teamActivity.teamID)
+		}
+
+		if len(a.userActivities[userID].teamActivities) == 0 {
+			a.userActivities[userID].state = entity.UserStatusDisConnected
+		}
+	}
+}
+
```
Comment (duanxianbo) [inline core/realtime/activity_store.go:44]: `ClientDisconnect`

**Pull request closed without merging**